Update readme with awesome-design-patterns

DovAmir opened the pull request:
Added link to new awesome list of software design patterns
https://github.com/DovAmir/awesome-design-patterns


<!-- Congrats on creating an Awesome list! 🎉 -->


<!-- Please fill in the below placeholders -->

**https://github.com/DovAmir/awesome-design-patterns**

**This is a list of design patterns, not only GOF but also architectual patterns, cloud patterns etc.. Many people are interested in patterns to quickly get to the gist of a subject but I havn't found any other list of design patterns resources.**



### By submitting this pull request I confirm I've read and complied with the below requirements 🖖

**Please read it multiple times. I spent a lot of time on these guidelines and most people miss a lot.**

## Requirements for your pull request

- I have read and understood the [contribution guidelines](https://github.com/sindresorhus/awesome/blob/master/contributing.md) and the [instructions for creating a list](https://github.com/sindresorhus/awesome/blob/master/create-list.md).
- This pull request has a descriptive title.<br>For example, `Add Name of List`, not `Update readme.md` or `Add awesome list`.
- The entry in the Awesome list should:
	- Include a short description about the project/theme of the list. **It should not describe the list itself.**<br>Example: `- [Fish](…) - User-friendly shell.`, not `- [Fish](…) - Resources for Fish.`.
	- Be added at the bottom of the appropriate category.
- The list I'm submitting complies with these requirements:


## Requirements for your Awesome list

- **Has been around for at least 30 days.**<br>That means 30 days from either the first real commit or when it was open-sourced. Whatever is most recent.
- It's the result of hard work and the best I could possibly produce.
- Non-generated Markdown file in a GitHub repo.
- **Includes a succinct description of the project/theme at the top of the readme.** [(Example)](https://github.com/willempienaar/awesome-quantified-self)
- The repo should have `awesome-list` & `awesome` as [GitHub topics](https://help.github.com/articles/about-topics). I encourage you to add more relevant topics.
- Not a duplicate.
- Only has awesome items. Awesome lists are curations of the best, not everything.
- Includes a project logo/illustration whenever possible.
	- Either fullwidth or placed at the top-right of the readme. [(Example)](https://github.com/sindresorhus/awesome-electron)
	- The image should link to the project website or any relevant website.
	- The image should be high-DPI. Set it to maximum half the width of the original image.
- Entries have a description, unless the title is descriptive enough by itself. It rarely is though.
- Includes the [Awesome badge](https://github.com/sindresorhus/awesome/blob/master/awesome.md#awesome-badge).
	- Should be placed on the right side of the readme heading.
	- Should link back to this list.
- Has a Table of Contents section.
	- Should be named `Contents`, not `Table of Contents`.
	- Should be the first section in the list.
	- Should only have one level of sub-lists, preferably none.
- Has an [appropriate license](https://github.com/sindresorhus/awesome/blob/master/awesome.md#choose-an-appropriate-license).
	- That means something like CC0, **not a code licence like MIT, BSD, Apache, etc.**
	- [WTFPL](http://www.wtfpl.net) and [Unlicense](http://unlicense.org) are not acceptable licenses.
	- If you use a license badge, it should be SVG, not PNG.
- Has [contribution guidelines](https://github.com/sindresorhus/awesome/blob/master/awesome.md#include-contribution-guidelines).
	- The file should be named `contributing.md`. Casing is up to you.
- Has consistent formatting and proper spelling/grammar.
	- The link and description are separated by a dash. <br>Example: `- [AVA](…) - JavaScript test runner.`
	- The description starts with an uppercase character and ends with a period.
	- Consistent and correct naming. For example, `Node.js`, not `NodeJS` or `node.js`.
- Doesn't include a Travis badge.<br>You can still use Travis for list linting, but the badge has no value in the readme.

Go to the top and read it again.

DovAmir commented:
that true, sorry about that. it has gained a huge following though.
so please accept the PR on the 4/7 .
thanks!

Solido commented:
When this repo will be older than 30 days it will have more stargazers than Vue.js ?! 
Very good content !

DovAmir commented:
This list has now  been around for 30 days.
this is a good time to accept the pull request

sindresorhus commented:
Hey, sorry for the lack of response, please see #1363.

mgramin commented:
```
$ awesome-lint

  README.md:1:1
  ×    1:1  Missing Awesome badge after the main heading                   awesome/badge
  ×   36:3  Found reference to undefined definition                        no-undefined-references
  ×   38:3  Found reference to undefined definition                        no-undefined-references
  ×   41:3  Found reference to undefined definition                        no-undefined-references
  ×   43:3  Found reference to undefined definition                        no-undefined-references
  ×   45:3  Found reference to undefined definition                        no-undefined-references
  ×   47:3  Found reference to undefined definition                        no-undefined-references
  ×   49:3  Found reference to undefined definition                        no-undefined-references
  ×   55:3  Found reference to undefined definition                        no-undefined-references
  ×   60:3  Found reference to undefined definition                        no-undefined-references
  ×   63:3  Found reference to undefined definition                        no-undefined-references
  ×   66:3  Found reference to undefined definition                        no-undefined-references
  ×   69:3  Found reference to undefined definition                        no-undefined-references
  ×   72:3  Found reference to undefined definition                        no-undefined-references
  ×   74:3  Found reference to undefined definition                        no-undefined-references
  ×   76:3  Found reference to undefined definition                        no-undefined-references
  ×   78:3  Found reference to undefined definition                        no-undefined-references
  ×   80:3  Found reference to undefined definition                        no-undefined-references
  ×   97:3  Remove trailing slash (http://cloudpatterns.org)               trailing-slash
  ×   98:3  Remove trailing slash (http://www.cloudcomputingpatterns.org)  trailing-slash
  ×   99:3  Remove trailing slash (https://gcp.solutions)                  trailing-slash
  ×  106:3  Remove trailing slash (http://microservicepatterns.org)        trailing-slash
  ×  109:3  Remove trailing slash (https://12factor.net)                   trailing-slash
  ×  119:3  Remove trailing slash (http://bigdatapatterns.org)             trailing-slash
  ×  124:3  Found reference to undefined definition                        no-undefined-references
  ×  131:3  Found reference to undefined definition                        no-undefined-references
  ×  146:3  Found reference to undefined definition                        no-undefined-references
  ×  148:3  Found reference to undefined definition                        no-undefined-references
  ×  155:3  Remove trailing slash (http://ui-patterns.com)                 trailing-slash
  ×  159:3  Found reference to undefined definition                        no-undefined-references

  30 errors
```

DovAmir commented:
here are some PR reviews I made
https://github.com/sindresorhus/awesome/pull/1180
https://github.com/sindresorhus/awesome/pull/1254
https://github.com/sindresorhus/awesome/pull/1267
https://github.com/sindresorhus/awesome/pull/1310
https://github.com/sindresorhus/awesome/pull/1336

DovAmir commented:
@sindresorhus  
I gave and received PR reviews.
Please consider this PR
Thanks!

DovAmir commented:
Hi @sindresorhus
anything stopping this from being merged?
Thanks!

sindresorhus commented:
```
❯ awesome-lint https://github.com/DovAmir/awesome-design-patterns
✖ Linting

  README.md:34:3
  ✖    1:1    Missing Awesome badge after the main heading                                                                           remark-lint:awesome/badge
  ✖   15:1    ToC item "GoF design patterns" does not match corresponding heading "Language/Framework design patterns (mainly GoF)"  remark-lint:awesome/toc
  ✖   34:3    Invalid list item link                                                                                                 remark-lint:awesome/list-item
  ✖   34:3    Found reference to undefined definition                                                                                remark-lint:no-undefined-references
  ✖   36:3    Invalid list item link                                                                                                 remark-lint:awesome/list-item
  ✖   36:3    Found reference to undefined definition                                                                                remark-lint:no-undefined-references
  ✖   39:3    Found reference to undefined definition                                                                                remark-lint:no-undefined-references
  ✖   39:3    Invalid list item link                                                                                                 remark-lint:awesome/list-item
  ✖   41:3    Found reference to undefined definition                                                                                remark-lint:no-undefined-references
  ✖   41:3    Invalid list item link                                                                                                 remark-lint:awesome/list-item
  ✖   42:4    Link to https://github.com/mlnv/design-patterns-for-hoomans is dead                                                    remark-lint:no-dead-urls
  ✖   43:3    Found reference to undefined definition                                                                                remark-lint:no-undefined-references
  ✖   43:3    Invalid list item link                                                                                                 remark-lint:awesome/list-item
  ✖   45:3    Found reference to undefined definition                                                                                remark-lint:no-undefined-references
  ✖   45:3    Invalid list item link                                                                                                 remark-lint:awesome/list-item
  ✖   47:3    Found reference to undefined definition                                                                                remark-lint:no-undefined-references
  ✖   47:3    Invalid list item link                                                                                                 remark-lint:awesome/list-item
  ✖   53:3    Found reference to undefined definition                                                                                remark-lint:no-undefined-references
  ✖   53:3    Invalid list item link                                                                                                 remark-lint:awesome/list-item
  ✖   58:3    Found reference to undefined definition                                                                                remark-lint:no-undefined-references
  ✖   58:3    Invalid list item link                                                                                                 remark-lint:awesome/list-item
  ✖   61:3    Found reference to undefined definition                                                                                remark-lint:no-undefined-references
  ✖   61:3    Invalid list item link                                                                                                 remark-lint:awesome/list-item
  ✖   64:3    Found reference to undefined definition                                                                                remark-lint:no-undefined-references
  ✖   64:3    Invalid list item link                                                                                                 remark-lint:awesome/list-item
  ✖   67:3    Found reference to undefined definition                                                                                remark-lint:no-undefined-references
  ✖   67:3    Invalid list item link                                                                                                 remark-lint:awesome/list-item
  ✖   70:3    Found reference to undefined definition                                                                                remark-lint:no-undefined-references
  ✖   70:3    Invalid list item link                                                                                                 remark-lint:awesome/list-item
  ✖   72:3    Found reference to undefined definition                                                                                remark-lint:no-undefined-references
  ✖   72:3    Invalid list item link                                                                                                 remark-lint:awesome/list-item
  ✖   74:3    Invalid list item link                                                                                                 remark-lint:awesome/list-item
  ✖   74:3    Found reference to undefined definition                                                                                remark-lint:no-undefined-references
  ✖   76:3    Invalid list item link                                                                                                 remark-lint:awesome/list-item
  ✖   76:3    Found reference to undefined definition                                                                                remark-lint:no-undefined-references
  ✖   78:3    Found reference to undefined definition                                                                                remark-lint:no-undefined-references
  ✖   78:3    Invalid list item link                                                                                                 remark-lint:awesome/list-item
  ✖  122:3    Invalid list item link                                                                                                 remark-lint:awesome/list-item
  ✖  122:3    Found reference to undefined definition                                                                                remark-lint:no-undefined-references
  ✖  129:3    Invalid list item link                                                                                                 remark-lint:awesome/list-item
  ✖  129:3    Found reference to undefined definition                                                                                remark-lint:no-undefined-references
  ✖  144:3    Found reference to undefined definition                                                                                remark-lint:no-undefined-references
  ✖  144:3    Invalid list item link                                                                                                 remark-lint:awesome/list-item
  ✖  146:3    Invalid list item link                                                                                                 remark-lint:awesome/list-item
  ✖  146:3    Found reference to undefined definition                                                                                remark-lint:no-undefined-references
  ✖  154:100  List item description must start with valid casing                                                                     remark-lint:awesome/list-item
  ✖  157:3    Invalid list item link                                                                                                 remark-lint:awesome/list-item
  ✖  157:3    Found reference to undefined definition                                                                                remark-lint:no-undefined-references

  48 errors

```

https://github.com/sindresorhus/awesome-lint

sindresorhus commented on readme.md:
```diff
@@ -595,6 +595,7 @@
 - [Open Source Maintainers](https://github.com/nayafia/awesome-maintainers#readme) - The experience of being an open source maintainer.
 - [Calculators](https://github.com/xxczaki/awesome-calculators#readme) - Calculators for every platform.
 - [Captcha](https://github.com/ZYSzys/awesome-captcha#readme) - A type of challenge–response test used in computing to determine whether or not the user is human.
+- [Design Patterns](https://github.com/DovAmir/awesome-design-patterns) - A curated list of software and architecture related design patterns.
```
> The entry in the Awesome list should:
* Include a short description about the project/theme of the list. **It should not describe the list itself.**

sindresorhus commented:
Drop the brackets `[` `]`.

sindresorhus commented:
I think the list should be renamed to `Awesome Software Design Patterns` so not to be confused with patterns in graphic design.

sindresorhus commented:
The list heading should be `Awesome design patterns` => `Awesome Design Patterns`

DovAmir commented:
I removed the [ ] 
changed the description
fixed the awesome-lint issues
changed the name to "Awesome Software Design Patterns"

anything else? @sindresorhus

sindresorhus commented on readme.md:
```diff
@@ -595,6 +595,7 @@
 - [Open Source Maintainers](https://github.com/nayafia/awesome-maintainers#readme) - The experience of being an open source maintainer.
 - [Calculators](https://github.com/xxczaki/awesome-calculators#readme) - Calculators for every platform.
 - [Captcha](https://github.com/ZYSzys/awesome-captcha#readme) - A type of challenge–response test used in computing to determine whether or not the user is human.
+- [Software Design Patterns](https://github.com/DovAmir/awesome-design-patterns) - Software and architecture related design patterns.
```
This should explain what "software design patterns" are, not what the list contains.

sindresorhus commented:
The items in https://github.com/DovAmir/awesome-design-patterns#programming-language-design-patterns need better titles and descriptions.

For example, `JavaScript - humans` doesn't say much about what to expect when clicking the link.

DovAmir commented:
added descriptions to all non trivial items in programming-language-design-patterns 
@sindresorhus 
I think this is now high quality and ready to be merged
thanks

sindresorhus commented:
https://github.com/sindresorhus/awesome/pull/1331/files#r272969143

sindresorhus commented:
Still many typos:

- `IOS` => `iOS`
- `[front-End Architecture]` => `Front-End Architecture]`
- `smiplifed` => `Simplified`
- `user Interface` => `User Interface`


---

```
❯ awesome-lint https://github.com/DovAmir/awesome-design-patterns
✖ Linting

  README.md:155:3
  ✖    1:1  License was not detected by GitHub                                                                                           remark-lint:awesome/github
  ✖    1:1  Missing Awesome badge after the main heading                                                                                 remark-lint:awesome/badge
  ✖   17:1  ToC item "Programming language design patterns" does not match corresponding heading "Programming Language Design Patterns"  remark-lint:awesome/toc
  ✖   95:3  Link to https://gcp.solutions is dead                                                                                        remark-lint:no-dead-urls
  ✖  155:3  Invalid list item link                                                                                                       remark-lint:awesome/list-item
  ✖  155:3  Found reference to undefined definition                                                                                      remark-lint:no-undefined-references
  ✖  160:3  Link to http://www.opensecurityarchitecture.org/cms/library/patternlandscape is dead                                         remark-lint:no-dead-urls

  7 errors
```

https://github.com/sindresorhus/awesome-lint

---

Descriptions should be mostly lower-case, not title-case. For example: `Strengths and Weaknesses of IoT Communication Patterns.` => `Strengths and weaknesses of IoT communication patterns.`

---

Generally, I feel the lists still needs more work.

I'm going to close this for now. I cannot keep going back and forth anymore. Try again in some months when you've had more time to work on it.
